File: node/src/transaction.test.ts
/*
 * Copyright 2020 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { CallOptions, Metadata, ServiceError, status } from '@grpc/grpc-js';
import { gateway as gatewayproto, peer } from '@hyperledger/fabric-protos';
import { CommitError } from './commiterror';
import { CommitStatusError } from './commitstatuserror';
import { Contract } from './contract';
import { Gateway, internalConnect, InternalConnectOptions } from './gateway';
import { Identity } from './identity/identity';
import { Network } from './network';
import { SubmitError } from './submiterror';
import { MockGatewayGrpcClient, newEndorseResponse } from './testutils.test';

describe('Transaction', () => {
    const expectedResult = 'TX_RESULT';
    const serviceError: ServiceError = Object.assign(new Error('ERROR_MESSAGE'), {
        code: status.UNAVAILABLE,
        details: 'DETAILS',
        metadata: new Metadata(),
    });

    let submitOptions: () => CallOptions;
    let commitStatusOptions: () => CallOptions;
    let client: MockGatewayGrpcClient;
    let identity: Identity;
    let signer: jest.Mock<Promise<Uint8Array>, Uint8Array[]>;
    let hash: jest.Mock<Uint8Array, Uint8Array[]>;
    let gateway: Gateway;
    let network: Network;
    let contract: Contract;

    beforeEach(() => {
        const now = new Date();
        const submitCallOptions = {
            deadline: now.setHours(now.getHours() + 1),
        };
        submitOptions = () => submitCallOptions; // Return a specific object to test modification
        const commitStatusCallOptions = {
            deadline: now.setHours(now.getHours() + 1),
        };
        commitStatusOptions = () => commitStatusCallOptions; // Return a specific object to test modification

        client = new MockGatewayGrpcClient();

        const endorseResponse = newEndorseResponse({
            result: Buffer.from(expectedResult),
        });
        client.mockEndorseResponse(endorseResponse);

        const commitResult = new gatewayproto.CommitStatusResponse();
        commitResult.setResult(peer.TxValidationCode.VALID);
        client.mockCommitStatusResponse(commitResult);

        identity = {
            mspId: 'MSP_ID',
            credentials: Buffer.from('CERTIFICATE'),
        };
        signer = jest.fn(undefined);
        signer.mockResolvedValue(Buffer.from('SIGNATURE'));
        hash = jest.fn(undefined);
        hash.mockReturnValue(Buffer.from('DIGEST'));

        const options: InternalConnectOptions = {
            identity,
            signer,
            hash,
            client,
            submitOptions,
            commitStatusOptions,
        };
        gateway = internalConnect(options);
        network = gateway.getNetwork('CHANNEL_NAME');
        contract = network.getContract('CHAINCODE_NAME');
    });

    it('throws on submit error', async () => {
        client.mockSubmitError(serviceError);
        const transaction = await contract.newProposal('TRANSACTION_NAME').endorse();
        const transactionId = transaction.getTransactionId();

        const t = transaction.submit();

        await expect(t).rejects.toThrow(SubmitError);
        await expect(t).rejects.toThrow(serviceError.message);
        await expect(t).rejects.toMatchObject({
            name: SubmitError.name,
            transactionId,
            code: serviceError.code,
            cause: serviceError,
        });
    });

    it('throws on commit status error', async () => {
        client.mockCommitStatusError(serviceError);
        const transaction = await contract.newProposal('TRANSACTION_NAME').endorse();
        const commit = await transaction.submit();
        const transactionId = commit.getTransactionId();

        const t = commit.getStatus();

        await expect(t).rejects.toThrow(CommitStatusError);
        await expect(t).rejects.toThrow(serviceError.message);
        await expect(t).rejects.toMatchObject({
            name: CommitStatusError.name,
            transactionId,
            code: serviceError.code,
            cause: serviceError,
        });
    });

    it('throws CommitError on commit failure', async () => {
        const commitResult = new gatewayproto.CommitStatusResponse();
        commitResult.setResult(peer.TxValidationCode.MVCC_READ_CONFLICT);
        client.mockCommitStatusResponse(commitResult);

        const t = contract.submitTransaction('TRANSACTION_NAME');

        await expect(t).rejects.toThrow(CommitError);
        await expect(t).rejects.toThrow('MVCC_READ_CONFLICT');
    });

    it('returns result', async () => {
        const result = await contract.submitTransaction('TRANSACTION_NAME');

        const actual = Buffer.from(result).toString();
        expect(actual).toBe(expectedResult);
    });

    it('sets endorsing orgs', async () => {
        await contract.submit('TRANSACTION_NAME', { endorsingOrganizations: ['org1', 'org3'] });
        const actualOrgs = client.getEndorseRequests()[0].getEndorsingOrganizationsList();
        expect(actualOrgs).toStrictEqual(['org1', 'org3']);
    });

    it('includes transaction ID in submit request', async () => {
        await contract.submitTransaction('TRANSACTION_NAME');

        const endorseRequest = client.getEndorseRequests()[0];
        const expected = endorseRequest.getTransactionId();

        const submitRequest = client.getSubmitRequests()[0];
        const actual = submitRequest.getTransactionId();

        expect(actual).toBe(expected);
    });

    it('uses signer for submit', async () => {
        signer.mockResolvedValue(Buffer.from('MY_SIGNATURE'));

        await contract.submitTransaction('TRANSACTION_NAME');

        const submitRequest = client.getSubmitRequests()[0];
        const signature = Buffer.from(submitRequest.getPreparedTransaction()?.getSignature_asU8() ?? '').toString();
        expect(signature).toBe('MY_SIGNATURE');
    });

    it('uses signer with newTransaction', async () => {
        signer.mockResolvedValue(Buffer.from('MY_SIGNATURE'));
        const unsignedProposal = contract.newProposal('TRANSACTION_NAME');
        const signedProposal = gateway.newProposal(unsignedProposal.getBytes());
        const unsignedTransaction = await signedProposal.endorse();

        const signedTransaction = gateway.newTransaction(unsignedTransaction.getBytes());
        await signedTransaction.submit();

        const submitRequest = client.getSubmitRequests()[0];
        const signature = Buffer.from(submitRequest.getPreparedTransaction()?.getSignature_asU8() ?? '').toString();
        expect(signature).toBe('MY_SIGNATURE');
    });

    it('uses signer for commit', async () => {
        signer.mockResolvedValue(Buffer.from('MY_SIGNATURE'));

        await contract.submitTransaction('TRANSACTION_NAME');

        const statusRequest = client.getCommitStatusRequests()[0];
        const signature = Buffer.from(statusRequest.getSignature()).toString();
        expect(signature).toBe('MY_SIGNATURE');
    });

    it('uses signer for newCommit', async () => {
        signer.mockResolvedValue(Buffer.from('MY_SIGNATURE'));

        const unsignedProposal = contract.newProposal('TRANSACTION_NAME');
        const signedProposal = gateway.newProposal(unsignedProposal.getBytes());
        const unsignedTransaction = await signedProposal.endorse();

        const signedTransaction = gateway.newTransaction(unsignedTransaction.getBytes());
        const unsignedCommit = await signedTransaction.submit();
        const deserializedSignedCommit = gateway.newCommit(unsignedCommit.getBytes());
        await deserializedSignedCommit.getStatus();

        const statusRequest = client.getCommitStatusRequests()[0];
        const signature = Buffer.from(statusRequest.getSignature()).toString();
        expect(signature).toBe('MY_SIGNATURE');
    });

    it('uses hash', async () => {
        hash.mockReturnValue(Buffer.from('MY_DIGEST'));

        await contract.submitTransaction('TRANSACTION_NAME');

        expect(signer).toHaveBeenCalledTimes(3); // endorse, submit and commit
        signer.mock.calls.forEach((call) => {
            const digest = call[0].toString();
            expect(digest).toBe('MY_DIGEST');
        });
    });

    it('commit returns transaction validation code', async () => {
        const commitResult = new gatewayproto.CommitStatusResponse();
        commitResult.setResult(peer.TxValidationCode.MVCC_READ_CONFLICT);
        client.mockCommitStatusResponse(commitResult);

        const commit = await contract.submitAsync('TRANSACTION_NAME');
        const status = await commit.getStatus();

        expect(status.code).toBe(peer.TxValidationCode.MVCC_READ_CONFLICT);
    });

    it('commit returns successful for successful transaction', async () => {
        const commitResult = new gatewayproto.CommitStatusResponse();
        commitResult.setResult(peer.TxValidationCode.VALID);
        client.mockCommitStatusResponse(commitResult);

        const commit = await contract.submitAsync('TRANSACTION_NAME');
        const status = await commit.getStatus();

        expect(status.successful).toBe(true);
    });

    it('commit returns unsuccessful for failed transaction', async () => {
        const commitResult = new gatewayproto.CommitStatusResponse();
        commitResult.setResult(peer.TxValidationCode.MVCC_READ_CONFLICT);
        client.mockCommitStatusResponse(commitResult);

        const commit = await contract.submitAsync('TRANSACTION_NAME');
        const status = await commit.getStatus();

        expect(status.successful).toBe(false);
    });

    it('commit returns block number', async () => {
        const commitResult = new gatewayproto.CommitStatusResponse();
        commitResult.setResult(peer.TxValidationCode.MVCC_READ_CONFLICT);
        commitResult.setBlockNumber(101);
        client.mockCommitStatusResponse(commitResult);

        const commit = await contract.submitAsync('TRANSACTION_NAME');
        const status = await commit.getStatus();

        expect(status.blockNumber).toBe(BigInt(101));
    });

    it('commit returns zero for missing block number', async () => {
        const commitResult = new gatewayproto.CommitStatusResponse();
        commitResult.setResult(peer.TxValidationCode.MVCC_READ_CONFLICT);
        client.mockCommitStatusResponse(commitResult);

        const commit = await contract.submitAsync('TRANSACTION_NAME');
        const status = await commit.getStatus();

        expect(status.blockNumber).toBe(BigInt(0));
    });

    it('submit uses specified call options', async () => {
        const deadline = Date.now() + 1000;
        const transaction = await contract.newProposal('TRANSACTION_NAME').endorse();

        await transaction.submit({ deadline });

        const actual = client.getSubmitOptions()[0];
        expect(actual.deadline).toBe(deadline);
    });

    it('submit uses default call options', async () => {
        const transaction = await contract.newProposal('TRANSACTION_NAME').endorse();

        await transaction.submit();

        const actual = client.getSubmitOptions()[0];
        expect(actual.deadline).toBe(submitOptions().deadline);
    });

    it('submit default call options are not modified', async () => {
        const expected = submitOptions().deadline;
        const deadline = Date.now() + 1000;
        const transaction = await contract.newProposal('TRANSACTION_NAME').endorse();

        await transaction.submit({ deadline });

        expect(submitOptions().deadline).toBe(expected);
    });

    it('commit uses specified call options', async () => {
        const deadline = Date.now() + 1000;
        const transaction = await contract.newProposal('TRANSACTION_NAME').endorse();
        const commit = await transaction.submit();

        await commit.getStatus({ deadline });

        const actual = client.getCommitStatusOptions()[0];
        expect(actual.deadline).toBe(deadline);
    });

    it('commit uses default call options', async () => {
        const transaction = await contract.newProposal('TRANSACTION_NAME').endorse();
        const commit = await transaction.submit();

        await commit.getStatus();

        const actual = client.getCommitStatusOptions()[0];
        expect(actual.deadline).toBe(commitStatusOptions().deadline);
    });

    it('commit default call options are not modified', async () => {
        const expected = commitStatusOptions().deadline;
        const deadline = Date.now() + 1000;
        const transaction = await contract.newProposal('TRANSACTION_NAME').endorse();
        const commit = await transaction.submit();

        await commit.getStatus({ deadline });

        expect(commitStatusOptions().deadline).toBe(expected);
    });
});

File: scenario/go/transaction_test.go
/*
Copyright 2021 IBM All Rights Reserved.

SPDX-License-Identifier: Apache-2.0
*/

package scenario

import (
	"fmt"

	"github.com/hyperledger/fabric-gateway/pkg/client"
	"github.com/hyperledger/fabric-gateway/pkg/identity"
	"github.com/hyperledger/fabric-protos-go-apiv2/gateway"
	"github.com/hyperledger/fabric-protos-go-apiv2/peer"
	"google.golang.org/grpc/status"
)

type Transaction struct {
	gateway     *client.Gateway
	contract    *client.Contract
	txType      TransactionType
	name        string
	options     []client.ProposalOption
	offlineSign identity.Sign
	result      []byte
	err         error
	blockNumber uint64
}

func NewTransaction(gateway *client.Gateway, contract *client.Contract, txType TransactionType, name string) *Transaction {
	return &Transaction{
		gateway:  gateway,
		contract: contract,
		txType:   txType,
		name:     name,
	}
}

func (transaction *Transaction) AddOptions(options ...client.ProposalOption) {
	transaction.options = append(transaction.options, options...)
}

func (transaction *Transaction) Invoke() error {
	switch transaction.txType {
	case Submit:
		transaction.result, transaction.err = transaction.submit()
	case Evaluate:
		transaction.result, transaction.err = transaction.evaluate()
	default:
		panic(fmt.Errorf("unknown transaction type: %v", transaction.txType))
	}

	return transaction.err
}

func (transaction *Transaction) SetOfflineSign(sign identity.Sign) {
	transaction.offlineSign = sign
}

func (transaction *Transaction) Result() []byte {
	return transaction.result
}

func (transaction *Transaction) Err() error {
	return transaction.err
}

func (transaction *Transaction) ErrDetails() []*gateway.ErrorDetail {
	statusErr := status.Convert(transaction.Err())
	var details []*gateway.ErrorDetail
	for _, detail := range statusErr.Details() {
		details = append(details, detail.(*gateway.ErrorDetail))
	}
	return details
}

func (transaction *Transaction) BlockNumber() uint64 {
	return transaction.blockNumber
}

func (transaction *Transaction) evaluate() ([]byte, error) {
	proposal, err := transaction.contract.NewProposal(transaction.name, transaction.options...)
	if err != nil {
		return nil, err
	}

	proposal, err = transaction.offlineSignProposal(proposal)
	if err != nil {
		return nil, err
	}

	return proposal.Evaluate()
}

func (transaction *Transaction) submit() ([]byte, error) {
	proposal, err := transaction.contract.NewProposal(transaction.name, transaction.options...)
	if err != nil {
		return nil, err
	}

	proposal, err = transaction.offlineSignProposal(proposal)
	if err != nil {
		return nil, err
	}

	unsignedTransaction, err := proposal.Endorse()
	if err != nil {
		return nil, err
	}

	signedTransaction, err := transaction.offlineSignTransaction(unsignedTransaction)
	if err != nil {
		return nil, err
	}

	result := signedTransaction.Result()

	unsignedCommit, err := signedTransaction.Submit()
	if err != nil {
		return result, err
	}

	signedCommit, err := transaction.offlineSignCommit(unsignedCommit)
	if err != nil {
		return result, err
	}

	status, err := signedCommit.Status()
	if err != nil {
		return result, err
	}

	transaction.blockNumber = status.BlockNumber

	if !status.Successful {
		return result, fmt.Errorf("commit failed with status %v (%v)", status.Code, peer.TxValidationCode_name[int32(status.Code)])
	}

	return result, nil
}

func (transaction *Transaction) offlineSignProposal(proposal *client.Proposal) (*client.Proposal, error) {
	if nil == transaction.offlineSign {
		return proposal, nil
	}

	bytes, signature, err := transaction.bytesAndSignature(proposal)
	if err != nil {
		return nil, err
	}

	proposal, err = transaction.gateway.NewSignedProposal(bytes, signature)
	if err != nil {
		return nil, err
	}

	return proposal, nil
}

func (transaction *Transaction) offlineSignTransaction(clientTransaction *client.Transaction) (*client.Transaction, error) {
	if nil == transaction.offlineSign {
		return clientTransaction, nil
	}

	bytes, signature, err := transaction.bytesAndSignature(clientTransaction)
	if err != nil {
		return nil, err
	}

	clientTransaction, err = transaction.gateway.NewSignedTransaction(bytes, signature)
	if err != nil {
		return nil, err
	}

	return clientTransaction, nil
}

func (transaction *Transaction) offlineSignCommit(commit *client.Commit) (*client.Commit, error) {
	if nil == transaction.offlineSign {
		return commit, nil
	}

	bytes, signature, err := transaction.bytesAndSignature(commit)
	if err != nil {
		return nil, err
	}

	commit, err = transaction.gateway.NewSignedCommit(bytes, signature)
	if err != nil {
		return nil, err
	}

	return commit, nil
}

type Signable interface {
	Digest() []byte
	Bytes() ([]byte, error)
}

func (transaction *Transaction) bytesAndSignature(signable Signable) ([]byte, []byte, error) {
	digest := signable.Digest()
	bytes, err := signable.Bytes()
	if err != nil {
		return nil, nil, err
	}

	signature, err := transaction.offlineSign(digest)
	if err != nil {
		return nil, nil, err
	}

	return bytes, signature, nil
}
